packages/tre/src/storage2/blob/store.ts

```diff
@@ -0,0 +1,191 @@
+import assert from "assert";
+import crypto from "crypto";
+import fs from "fs/promises";
+import path from "path";
+import { arrayBuffer } from "stream/consumers";
+import { ReadableStream } from "stream/web";
+import {
+  InclusiveRange,
+  MultipartOptions,
+  MultipartReadableStream,
+  createArrayReadableStream,
+  createFileReadableStream,
+  createFileWritableStream,
+  createMultipartArrayReadableStream,
+  createMultipartFileReadableStream,
+} from "./streams";
+
+const MODE_READ_ONLY = 0o444;
+
+// Serialisable, opaque, unguessable blob identifier
+export type BlobId = string;
+
+export interface BlobStore {
+  // Database for binary large objects. Provides single and multi-ranged
+  // streaming reads and writes.
+  //
+  // Blobs have unguessable identifiers, can be deleted, but are otherwise
+  // immutable. These properties make it possible to perform atomic updates with
+  // the SQLite metadata store. No other operations will be able to interact
+  // with the blob until it's committed to the metadata store, because they
+  // won't be able to guess the ID, and we don't allow listing blobs.
+  //
+  // For example, if we put a blob in the store, then fail to insert the blob ID
+  // into the SQLite database for some reason during a transaction (e.g.
+  // `onlyIf` condition failed), no other operations can read that blob because
+  // the ID is lost (we'll just background-delete the blob in this case).
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+
+  put(stream: ReadableStream<Uint8Array>): Promise<BlobId>;
+
+  delete(id: BlobId): Promise<void>;
+}
+
+function generateBlobId(): BlobId {
```

Comment:
What about just using a UUIDv4? Or a ULID if the time component is important?

Reply:
Not remotely blocking for this PR, but I _think_ `crypto.randomUUID()` should be a dependency-free way to get a UUID